### PR TITLE
Quaternion/new kinematics cleanup

### PR DESCRIPTION
**Pull request by tkoolen**
See https://github.com/RobotLocomotion/drake/pull/716 and https://github.com/mitdrc/drc/issues/934
* forwardKin options struct
* modify dTransformAdjoint and turn it into dTransformSpatialForce for more convenience and efficiency
* get rid of forwardKinV, instead merge forwardKinV into forwardKin,use in_terms_of_v boolean in options struct
* add in_terms_of_qdot option to geometricJacobian
* update dcross Matlab implementation: add da and db inputs.
* differentiate between nq and nv in C++ code (addition w.r.t. https://github.com/mitdrc/drc/issues/934)
* made eigenToMatlab more efficient and moved it and related functions to drakeUtil. (addition w.r.t. https://github.com/mitdrc/drc/issues/934)

### CODE REVIEW

File: systems/controllers/controlUtil.cpp [hunk 1]
```diff
@@ -42,15 +42,6 @@ void angleDiff(MatrixBase<DerivedPhi1> const &phi1, MatrixBase<DerivedPhi2> cons
...
 
-template <typename DerivedA>
-mxArray* eigenToMatlab(const DerivedA &m)
-{
- mxArray* pm = mxCreateDoubleMatrix(static_cast<int>(m.rows()),static_cast<int>(m.cols()),mxREAL);
- if (m.rows()*m.cols()>0)
-   memcpy(mxGetPr(pm),m.data(),sizeof(double)*m.rows()*m.cols());
- return pm;
-}
-
```
Comment (tkoolen) [inline systems/controllers/controlUtil.cpp]: moved to drakeUtil.

File: systems/plants/@RigidBodyManipulator/doKinematics.m [hunk 2]
```diff
@@ -254,15 +254,15 @@
...
   if options.compute_gradients
-    [kinsol.qdotToV, kinsol.dqdotToVdq] = qdotToV(model, q);
-    [kinsol.vToqdot, kinsol.dvToqdotdq] = vToqdot(model, q);
+    [kinsol.qdotToV, kinsol.dqdotToVdq] = computeQdotToV(bodies, q);
+    [kinsol.vToqdot, kinsol.dvToqdotdq] = computeVToqdot(bodies, q);
   else
-    kinsol.qdotToV = qdotToV(model, q);
-    kinsol.vToqdot = vToqdot(model, q);
+    kinsol.qdotToV = computeQdotToV(bodies, q);
+    kinsol.vToqdot = computeVToqdot(bodies, q);
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: Decided to change kinsol.qdotToV and kinsol.vToQdot to a cell array of matrices instead of one big matrix.

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m [hunk 3]
```diff
@@ -1,30 +1,36 @@
-function [J, v_indices, dJ] = geometricJacobian(obj, kinsol, base, end_effector, expressed_in)
-%GEOMETRICJACOBIAN Computes the geometric Jacobian from base to endEffector
+function [J, v_or_qdot_indices, dJ] = geometricJacobian(obj, kinsol, base, end_effector, expressed_in, in_terms_of_qdot)
+%GEOMETRICJACOBIAN Computes the geometric Jacobian from base to end_effector
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: or should I use an options struct here as well?

File: systems/plants/RigidBodyManipulator.cpp [hunk 4]
```diff
@@ -338,7 +338,7 @@ void RigidBodyManipulator::resize(int ndof, int num_featherstone_bodies, int num
   bodies.reserve(num_bodies);
   for (int i = last_num_bodies; i < num_bodies; i++) {
     bodies.push_back(std::unique_ptr<RigidBody>(new RigidBody()));
-    bodies[i]->dofnum = i - 1;
+//    bodies[i]->dofnum = i - 1;
   } // setup default dofnums
```
Comment (tkoolen) [inline systems/plants/RigidBodyManipulator.cpp]: I didn't see why this was necessary.

File: systems/plants/URDFRigidBodyManipulator.cpp [hunk 5]
```diff
@@ -337,6 +339,7 @@ bool URDFRigidBodyManipulator::addURDF(boost::shared_ptr<urdf::ModelInterface> _
       // pitch is irrelevant
       bodies[index]->Ttree = Matrix4d::Identity();
       bodies[index]->T_body_to_joint = Matrix4d::Identity();
+      bodies[index]->setJoint(std::unique_ptr<RollPitchYawFloatingJoint>(new RollPitchYawFloatingJoint(jointname, Isometry3d::Identity())));
```
Comment (tkoolen) [inline systems/plants/URDFRigidBodyManipulator.cpp]: Previously I wasn't making a DrakeJoint for bodies with an rpy parameterized floating joint.

File: systems/plants/test/evaluateNewSecondDerivativeMethod.m [hunk 6]
```diff
@@ -0,0 +1,75 @@
+function evaluateNewSecondDerivativeMethod()
+%NOTEST
```
Comment (tkoolen) [inline systems/plants/test/evaluateNewSecondDerivativeMethod.m]: I'm trying a different method to compute the gradient of a geometric Jacobian. It works and may be faster. Will pursue this later. Is it OK if I commit this?

File: systems/plants/@RigidBodyManipulator/doKinematics.m [hunk 7]
```diff
@@ -52,7 +52,7 @@
     valuecheck(b_compute_second_derivatives, []);
```
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: I know these are not from you, but valuechecks here seem superfluous (and potentially expensive), especially given the logic you've now added below.

File: systems/plants/@RigidBodyManipulator/forwardKin.m [hunk 8]
```diff
@@ -1,4 +1,4 @@
-function [x,J,dJ] = forwardKin(obj,kinsol,body_or_frame_ind,pts,rotation_type, base_or_frame_ind)
+function [x,J,dJ] = forwardKin(obj,kinsol,body_or_frame_ind,pts,rotation_type, base_or_frame_ind, options)
```
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/forwardKin.m]: since we're pulling the trigger on the options structure, should we move rotation_type and base_or_frame_ind into that, too?  (also, could we make it base_or_frame_id)

File: systems/plants/@RigidBodyManipulator/forwardKin.m [hunk 9]
```diff
@@ -32,27 +32,41 @@
...
 
+if nargin < 7
+  options = struct();
+  if nargin > 4
+    options.rotation_type = rotation_type;
+  end
+else
+  % if options struct is passed in, make sure that rotation_type argument
+  % is empty to avoid conflicts
+  valuecheck(rotation_type, []);
```
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/forwardKin.m]: good.  i'd say let's zap the rotation_type input (and definitely this valuecheck)

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m (hunk 3 above)
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: my vote would be for expressed_in and in_terms_of_qdot to be in an options struct.

File: systems/plants/RigidBody.h [hunk 10]
```diff
@@ -45,7 +45,8 @@ class DLLEXPORT_RBM RigidBody {
   static const std::set<int> defaultRobotNumSet;
 // note: it's very ugly, but parent,dofnum,and pitch also exist currently (independently) at the rigidbodymanipulator level to represent the featherstone structure.  this version is for the kinematics.
   int parent;
-  int dofnum;
+  int dofnum; // interpreted as start of position_num from Matlab
```
Comment (RussTedrake) [inline systems/plants/RigidBody.h]: should we make it consistent, then.  position_num_start?

File: systems/plants/RigidBodyManipulator.h [hunk 11]
```diff
@@ -179,6 +179,7 @@ class DLLEXPORT_RBM RigidBodyManipulator
   std::vector<std::string> robot_name;
 
   int num_dof; // treated as nq now; TODO: rename to nq
```
Comment (RussTedrake) [inline systems/plants/RigidBodyManipulator.h]: rename to num_positions?

File: systems/plants/test/evaluateNewSecondDerivativeMethod.m (hunk 6 above)
Comment (RussTedrake) [inline systems/plants/test/evaluateNewSecondDerivativeMethod.m]: i think it's fine to commit since it's in test.  we also sometimes have dev subdirs.

File: util/drakeUtil.h [hunk 12]
```diff
@@ -45,6 +46,45 @@ template <typename Derived> inline void destroyDrakeMexPointer(const mxArray* mx
 //  mexPrintf(mexIsLocked() ? "mex file is locked\n" : "mex file is unlocked\n");
 }
 
+template <typename DerivedA>
+mxArray* eigenToMatlab(const DerivedA &m)
```
Comment (RussTedrake) [inline util/drakeUtil.h]: in general, we don't want to have templated method implementations in header files.  in this case, since it's almost a macro (no cross dependence between different object files), it *might* be ok?

File: systems/plants/@RigidBodyManipulator/forwardKin.m (hunk 8 above)
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/forwardKin.m]: rotation_type is in the options structure, but I didn't want to change existing functionality yet. If you think we should I can change it.
Should I also change body_or_frame_ind to body_or_frame_id then?
My thoughts were to have base_or_frame_ind not be an option but an actual argument just like body_or_frame_ind. I'm not advocating this, but why not move body_or_frame_ind to the options struct too then?

File: systems/plants/@RigidBodyManipulator/forwardKin.m (hunk 9 above)
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/forwardKin.m]: OK.

File: systems/plants/RigidBody.h (hunk 10 above)
Comment (tkoolen) [inline systems/plants/RigidBody.h]: Yes, but my reason for leaving it like this for now is to not cause merge conflicts all over the place. Maybe it'll be easier to do a separate pull request where we rename all of these things right after the DRC meeting so that people won't be affected as much.

File: systems/plants/RigidBodyManipulator.h (hunk 11 above)
Comment (tkoolen) [inline systems/plants/RigidBodyManipulator.h]: Yeah, that's better.

File: systems/plants/@RigidBodyManipulator/forwardKin.m (hunk 8 above)
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/forwardKin.m]: My suggestion would be to change to the new format, with options coming right after points.  To transition, we should check if the options is numeric (instead of a struct); if it is then we warnOnce and then treat it as if it was rotation_type.

In my view the mandatory arguments are kinsol, body_or_frame_id, and pts...  (it wouldn't make sense to without any of those) everything else as an option?

File: util/drakeUtil.h (hunk 12 above)
Comment (tkoolen) [inline util/drakeUtil.h]: Well I suppose you know my opinion on this by now. I can do the explicit instantiation thing again, but it just eats up my time with nothing to show for it in my view.

Comment (RussTedrake) [inline util/drakeUtil.h]: Yes, I know your view on it.  :)  I think it's ok in this case, simply because this function doesn't do any non-trivial sharing between separately compiled objects.

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m (hunk 3 above)
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: I think I actually prefer it as it is for now, as I don't see any other options being added in the near future, and the expressed_in and in_terms_of_qdot arguments are often needed.

**Comment by tkoolen** [comment]
I made changes to the method signatures of doKinematics and forwardKin, but I've turned the warnings for using the old method signatures off for now. There are 322 doKinematics calls and 370 forwardKin calls in Drake, so I'd be touching a lot of files. I would really prefer doing that after my next pull request (manipulatorDynamics).
So if you're OK with it I'd like to leave the following as to dos to be resolved after this pull request:
- [ ] fix doKinematics calls, turn on warning in doKinematics
- [ ] fix forwardKin calls, turn on warning in forwardKin
- [ ] RigidBody.dofnum --> position_num_start
- [ ] RigidBodyManipulator.dofnum --> num_positions

**Comment by RussTedrake** [comment]
Sounds ok to me.

**Comment by tkoolen** [comment]
Are you still waiting for a code update? I think I made all the required changes for this pull request.

File: systems/plants/@RigidBodyManipulator/doKinematics.m [hunk 13]
```diff
@@ -10,64 +10,57 @@
...
+
+% method signature transition
+if nargin > 3
+  options_copy = options;
+end
+warn_signature_changed = false;
+if nargin == 5
+  warn_signature_changed = true;
+  options = struct();
+  options.compute_gradients = v;
```
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: compute_gradients is a confusing name for this.  we always compute first derivatives, and this is only for second derivatives.. correct?

**Comment by RussTedrake** [comment]
let me know what you think about compute_gradients.  otherwise, i think it looks ready to go.

File: systems/plants/@RigidBodyManipulator/doKinematics.m (hunk 13 above)
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: It depends on what you view as a gradient and whether it's about computing gradients of things inside doKinematics or the ability to compute gradients of things using the results of the doKinematics call.
I named the option compute_gradients in doKinematics because when it's set to true, doKinematics computes gradients of everything that would normally be put in kinsol (what I consider the base level).
The base level allows you to compute geometric Jacobians, which aren't really gradients of anything, and also forwardKin Jacobians, which are. But I also came up with an algorithm (currently not in use) that allows you to compute gradients of geometric Jacobians using only the information in a kinsol from a non-gradient doKinematics call. A non-gradient kinsol also allows you to compute a mass matrix, which can be considered a base level thing, or the Hessian of kinetic energy w.r.t. velocity. I know that's kind of a stretch, but you see my point that if you consider the ability to compute gradients of things outside of doKinematics, it's not 100% clear anymore.
Any suggestions for what the name should be if you still think it's confusing?

Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: this points out the clear need to document the options structure at the top of the method.  :)

the dTdq, etc (and even the analytic jacobians) are clearly gradients which are computed right now, even when compute_gradients=false, so i think the name has to change. 
do all of these additional things that get build co-exist?  would it make sense to have multiple options for computing second derivatives, for computing geometry jac grads, etc.?  if a single option is still appropriate, then i'd be ok with even a `compute_additional_gradients` if it was well documented.

Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: Yes, I need to add documentation.

> would it make sense to have multiple options for computing second derivatives, for computing geometry jac grads, etc.

I think the current options reflect the typical usage cases:
1. compute_gradients = false, compute_JdotV = false: position control, planning using centroidal dynamics and full kinematics
2. compute_gradients = false, compute_JdotV = true: model based control, QP controller (requires Jdot * v terms in Cartesian constraints, dynamics bias term (C), CMMdot * v)
3. compute_gradients = true: second order methods, 'Michael's stuff' is what I hear

I'm OK with compute_additional_gradients, although I still think compute_gradients is clearer since that's exactly what you're telling doKinematics to do.

dTdq does not get computed in doKinematics when compute_gradients = false. As an aside, you could compute a dTdq for a single body from information using a kinsol with compute_gradients = false with `dHomogTrans` based on the S (motion subspace) and qdot_to_v fields of kinsol, but dTdq for all bodies is not precomputed in doKinematics when compute_gradients = false.

Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: OK. I guess we can call it compute_gradients, so long as it is clearly documented that jacobians are computed even when compute_gradients = false.

**Comment by tkoolen** [comment]
Just pushed the updated documentation.
